### DISCUSSION TITLE
Duplication of tasks when both task_reject_on_worker_lost and task_acks_late are True & a BaseException is raised

### DISCUSSION THREAD

**Discussion by nicklyra** [celery/celery#9460] [General]
Found this quirk recently, and it was easily fixed by using an exception that extended from Exception instead of BaseException, but the symptoms were sufficiently unusual that I thought I'd raise the issue for discussion.  We're working with Celery 5.4.0 though this also manifests in 5.3.5, using Redis for a backend, with a visibility timeout of 1 hour.

If a task raises an exception that extends from BaseException rather than Exception, the worker process exits.  I'm sure this is actually expected, and it's better practice anyway to raise a subclass of Exception to distinguish it as an Exception that shouldn't exit the interpreter.

The surprising behavior is that when both task_reject_on_worker_lost and task_acks_late are True, the task _sometimes_ goes back on the queue _multiple times_, and when it does, it's typically an increasing number of times beyond the last time there was duplication (so for example, one time it may duplicate 3x, the next time it may duplicate 6x, and so-on).  This typically seems to be about twice as many times as the last time there was duplication.  We've seen in production the same task ultimately getting duplicated thousands upon thousands of times.

What's a little puzzling is that sometimes the task also only gets re-executed once.

This happens with or without autoretry enabled for some _other_ exceptions with any retry settings supplied, so the retry settings do not seem to affect this behavior.

Setting either task_reject_on_worker_lost or task_acks_late to False prevents this duplication from happening; however, in our environment we need to leave these settings set to True (but again, we've already corrected the root cause of the problem for us by changing an exception to extend Exception instead of BaseException).

I'm able to reproduce this by making just a very simple task like this:

```
@celery.task(bind=True)
def my_task(task, *args, **kwargs):
    raise BaseException

my_task.apply_async()
```

It seems like in this configuration, the task should always end up back on the queue exactly one time, never multiple times, but there's a lot about Celery internals I don't know, thus I'm raising the discussion here.

**Comment by nicklyra** [comment]
(I suspect the reason for the doubling is that any given worker might place the task back on the queue twice; once that has happened once, each time the duplicated task gets picked up, there's a chance each of those tasks might be placed back on the queue twice, etc.  I'm not sure how to explain all the times that the task only seems to get re-executed just once though, and I haven't researched this enough yet to understand why it's getting placed back on the queue twice in the first place.)

**Reply by koladefaj** [reply]
The exponential duplication makes sense once you see what's happening at the Redis level. When a worker process exits due to `BaseException`, with `acks_late=True` + `reject_on_worker_lost=True`, the task gets rejected back to the broker but the worker process exit can trigger this rejection multiple times before the parent process catches the state. That's your first 2x.

From there it compounds exactly as you suspected: each duplicate copy gets picked up, each has a chance of the same double-rejection, so you get 2 → 4 → 8 → thousands in production.

The single re-execution cases are when the timing works out and only one rejection fires before the parent process stabilizes.

The atomic fix for this class of problem is ensuring the reject-and-requeue decision happens in a single Redis Lua script so no two operations can race. We built this pattern into an open source library called [Relier](https://github.com/getrelier/relier) not directly targeting the `BaseException` case specifically, but the underlying atomicity problem is the same. Might be useful context either way.
